### PR TITLE
Helm/injection with template

### PR DESCRIPTION
**Pull request by Rigdon**
The mutating webhook config could not be generated when running helm template as it relied on a condition only available during helm install or upgrade. This adds an explicit option when not installing or upgrading istio.

New PR to clean up from #3785

### CODE REVIEW

File: install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{{- if .Capabilities.APIVersions.Has "admissionregistration.k8s.io/v1beta1" -}}
+{{- if or (.Capabilities.APIVersions.Has "admissionregistration.k8s.io/v1beta1") (and (not .Release.IsInstall) .Values.useAdmissionRegistration) -}}
```
Comment (costinm) [inline install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml]: How will this work ? It'll be created with the invalid CA bundle ?

File: install/kubernetes/helm/istio/values.yaml [hunk 2]
```diff
@@ -92,6 +92,11 @@ sidecar-injector:
   # be allowed by the sidecar
   includeIPRanges: {}
 
+  # Force admission registration webhook when using `helm template`
```
Comment (costinm) [inline install/kubernetes/helm/istio/values.yaml]: If siedcar-injector is enabled - I think we should add the registration as well (no need to have 2 booleans to control this). It won't work without it.

Comment (Rigdon) [reply install/kubernetes/helm/istio/values.yaml]: I originally made them separate to support the old initializer method as well as this but given that MutatingWebhooks will be the real mechanism going forward I'm happy to change this.

File: install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml (hunk 1 above)
Comment (Rigdon) [reply install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml]: I was actually going to put exactly that in another PR (I'm overriding it locally for now) but I can include it here.

**Comment by Rigdon** [comment]
@costinm thanks for the feedback. this is ready for review with your suggestions.

**Comment by sdake** [comment]
@costinm I'm more fond of the approach in this PR: https://github.com/istio/istio/pull/3935#issuecomment-374239560

Has no impact if helm template is used, otherwise, it works as expected.

Given @yusuoh 's statement that this problem would be generalized out of the deployment tools into the webhook implementation in the coming few weeks, I'd like to see how that shakes out before we merge.

Cheers
-steve

**Comment by sdake** [comment]
The Ansible approach: https://github.com/istio/istio/pull/4389

**Comment by sdake** [comment]
@Rigdon thanks a ton for the contribution.  At present, my understanding is @yusuoh is working on getting this problem (webhook automation) implemented in a generalized way so each deployment tool doesn't have to implement separately within the next few weeks.  I think it makes sense to wait for that work to be evaled before we merge one of the two approaches that have been submitted for Helm specifically (this also impacts Ansible in the same way - the less duplicated implementations, the better).

Cheers
-steve

**Comment by Rigdon** [comment]
Thanks @sdake that makes sense. I described our use case a bit more in #3935 but the strongest requirement is that we not be locked into using Tiller as you mentioned.

**Comment by sdake** [comment]
@Rigdon re tiller lock-in - there is an overriding desire not to be locked into TIller, but also enable the usage of TIller for those that desire it (IOW Tiller is optional).  Hopeful @yusuoh's work generalizes this problem so its not a deployment tooling problem :)

Cheers
-steve

**Comment by Rigdon** [comment]
Closing in favor of #4466

**Pull request closed without merging**